### DISCUSSION TITLE
Are Linq operators with a generic state parameter considered?

### DISCUSSION THREAD

**Discussion by hesselmonk** [dotnet/csharplang#7672] [General] [answered]
I am a big fan of the declarative nature of Linq statements, but it is hard to use without causing (accidental) heap allocations. With an added state parameter a lot of those allocations can be avoided because the delegate instance can be cached.

In this example, the `N` method allocates on each call, while the `O` method caches the delegate: 

```cs
using System;
using System.Collections.Generic;
using System.Linq;

public class C 
{
    public int Threshold;
    
    public bool M(List<int> list) 
    {
        return list.Any(i => i > 42);
    }
    
    public bool N(List<int> list) 
    {
        return list.Any(i => i > Threshold);
    }
    
    public bool O(List<int> list) 
    {
        return list.Any(this, (i, s) => i > s.Threshold);
    }
    
}

public static class IEnumerableExt
{
    public static bool Any<T, TState>(this IEnumerable<T> e, TState state, Func<T, TState, bool> predicate)
    {
        foreach (var item in e)
        {
            if (predicate.Invoke(item, state)) return true;
        }
        
        return false;
    }
}
```

Are these extensions to the Linq operators considered? Or are there other improvements to the allocations incurred by Linq in the works?

**Comment by 333fred** [comment, marked as answer]
Are you asking for new extension methods, or new query syntax? If the former, you'll want to open an API proposal in the dotnet/runtime repo.

**Reply by hesselmonk** [reply]
Sorry, yes, I am more interested in the API than new query syntax. I'll move my question to the dotnet/runtime repo.
